commit.go

```diff
@@ -0,0 +1,432 @@
+			}
+		}()
+	}
+
+	var (
+		cs     = client.ContentStore()
+		differ = client.DiffService()
+		snName = info.Snapshotter
+		sn     = client.SnapshotService(snName)
+	)
+
+	// Don't gc me and clean the dirty data after 1 hour!
+	ctx, done, err := client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
+	if err != nil {
+		return emptyDigest, errors.Wrapf(err, "failed to create lease for commit")
+	}
+	defer done(ctx)
+
+	diffLayerDesc, diffID, err := createDiff(ctx, id, sn, cs, differ)
+	if err != nil {
+		return emptyDigest, errors.Wrap(err, "failed to export layer")
+	}
+
+	imageConfig, err := generateCommitImageConfig(ctx, container, diffID, opts)
+	if err != nil {
+		return emptyDigest, errors.Wrap(err, "failed to generate commit image config")
+	}
+
+	rootfsID := identity.ChainID(imageConfig.RootFS.DiffIDs).String()
+	if err := applyDiffLayer(ctx, rootfsID, baseImgConfig, sn, differ, diffLayerDesc); err != nil {
+		return emptyDigest, errors.Wrap(err, "failed to apply diff")
+	}
+
+	commitManifestDesc, configDigest, err := writeContentsForImage(ctx, cs, snName, baseImg, imageConfig, diffLayerDesc)
+	if err != nil {
+		return emptyDigest, err
+	}
+
+	// image create
+	img := images.Image{
+		Name:      opts.ref,
+		Target:    commitManifestDesc,
+		CreatedAt: time.Now(),
+	}
+
+	if _, err := client.ImageService().Update(ctx, img); err != nil {
+		if !errdefs.IsNotFound(err) {
+			return emptyDigest, err
+		}
+
+		if _, err := client.ImageService().Create(ctx, img); err != nil {
+			return emptyDigest, errors.Wrapf(err, "failed to create new image %s", opts.ref)
+		}
+	}
+	return configDigest, nil
+}
+
+// generateCommitImageConfig returns commit oci image config based on the container's image.
+func generateCommitImageConfig(ctx context.Context, container containerd.Container, diffID digest.Digest, opts *commitOpts) (ocispec.Image, error) {
+	spec, err := container.Spec(ctx)
+	if err != nil {
+		return ocispec.Image{}, err
+	}
+
+	img, err := container.Image(ctx)
+	if err != nil {
+		return ocispec.Image{}, err
+	}
+
+	baseConfig, err := readImageConfig(ctx, img)
+	if err != nil {
+		return ocispec.Image{}, err
+	}
+
+	if opts.author == "" {
+		opts.author = baseConfig.Author
+	}
+
+	createdBy := ""
+	if spec.Process != nil {
+		createdBy = strings.Join(spec.Process.Args, " ")
+	}
+
+	createdTime := time.Now()
+	return ocispec.Image{
+		Architecture: runtime.GOARCH,
+		OS:           runtime.GOOS,
+		Created:      &createdTime,
+		Author:       opts.author,
+		Config:       baseConfig.Config, // TODO(fuweid): how to update the USER/ENV/CMD/... fields?
+		RootFS: ocispec.RootFS{
+			Type:    "layers",
+			DiffIDs: append(baseConfig.RootFS.DiffIDs, diffID),
+		},
+		History: append(baseConfig.History, ocispec.History{
+			Created:    &createdTime,
+			CreatedBy:  createdBy,
+			Author:     opts.author,
+			Comment:    opts.message,
+			EmptyLayer: (diffID == emptyGZLayer),
```

Comment:
This is not robust, but I can't come up with a better idea.

Reply:
Yeah. It is kind of hacky. 😂